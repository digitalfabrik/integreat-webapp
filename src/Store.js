import { applyMiddleware, combineReducers, compose, createStore } from 'redux'

import thunkMiddleware from 'redux-thunk'
import { initializeCurrentLocation, routerForBrowser } from 'redux-little-router'
import createBrowserHistory from 'history/createBrowserHistory'

import routes from 'routes'
import endpointReducers from 'modules/endpoint/reducers'
import { createLogger } from 'redux-logger'

import setLanguageChangeUrlsReducer from 'modules/language/reducers/setLanguageChangeUrls'

class Store {
  init (initialState) {
    this._history = createBrowserHistory()

    this.history.listen((location, action) => {
      // Keep default behavior of restoring scroll position when user:
      // - clicked back button
      // - clicked on a link that programmatically calls `history.goBack()`
      // - manually changed the URL in the address bar (here we might want
      // to scroll to top, but we can't differentiate it from the others)
      if (action === 'POP') {
        return
      }
      // In all other cases, scroll to top
      window.scrollTo(0, 0)
    })

    // Additional reducers
    const {
      enhancer,
      reducer,
      middleware
    } = routerForBrowser({routes, basename: '', history: this.history})

    /**
     * The middlewares of this app, add additional middlewares here
     * @type {[*]}
     */
    let middlewares = [
      middleware,
      thunkMiddleware // Allows to return functions in actions
    ]

    // eslint-disable-next-line no-undef
    if (__DEV__) {
      middlewares.push(createLogger()) // Logs all state changes in console
    }

    /**
     * Configures the main store which holds the global state of the app
     *
     * @param preloadedState
     * @returns {*} A configured store
     */
    const configureStore = function configureStore (preloadedState) {
      return createStore(
        combineReducers({
          ...endpointReducers,
          router: reducer,
          languageChangeUrls: setLanguageChangeUrlsReducer
        }),
        preloadedState,
        compose(enhancer, applyMiddleware(...middlewares))
      )
    }

    this._store = configureStore(initialState)

    const initialLocation = this.getState().router

    if (initialLocation) {
      this.dispatch(initializeCurrentLocation(initialLocation))
    }
  }

  /**
   * @returns Gets redux state
   */
  getState () {
    return this._store.getState()
  }

  get redux () {
    return this._store
  }

  dispatch (action) {
    return this._store.dispatch(action)
  }

  subscribe (fn) {
    return this._store.subscribe(fn)
  }

  /**
   * @returns Gets the current history implementation
   */
  get history () {
    return this._history
  }
}

export default Store
