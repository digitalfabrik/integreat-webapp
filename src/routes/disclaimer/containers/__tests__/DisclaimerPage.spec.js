import React from 'react'
import { mount, shallow } from 'enzyme'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import ConnectedDisclaimerPage, { DisclaimerPage } from '../DisclaimerPage'
import LanguageModel from 'modules/endpoint/models/LanguageModel'
import DisclaimerModel from 'modules/endpoint/models/DisclaimerModel'
import Payload from 'modules/endpoint/Payload'
import Store from '../../../../Store'

const location = 'augsburg'
const languages = [
  new LanguageModel('en', 'English'),
  new LanguageModel('de', 'Deutsch'),
  new LanguageModel('ar', 'Arabic')
]

const disclaimer = new DisclaimerModel({
  id: 1689, title: 'Feedback, Kontakt und mÃ¶gliches Engagement', content: 'this is a test content'
})

describe('DisclaimerPage', () => {
  const mockSetLanguageChangeUrls = jest.fn()

  test('should render', () => {
    const wrapper = shallow(
      <DisclaimerPage languages={languages}
                      location={location}
                      disclaimer={disclaimer}
                      setLanguageChangeUrls={mockSetLanguageChangeUrls} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should dispatch once in componentDidMount', () => {
    const mockSetLanguageChangeUrls = jest.fn()

    const disclaimerPage = shallow(
      <DisclaimerPage languages={languages}
                      location={location}
                      disclaimer={disclaimer}
                      setLanguageChangeUrls={mockSetLanguageChangeUrls} />
    ).instance()

    expect(mockSetLanguageChangeUrls.mock.calls).toHaveLength(1)
    expect(mockSetLanguageChangeUrls).toBeCalledWith(disclaimerPage.mapLanguageToUrl, languages)
  })

  test('mapLanguageToUrl', () => {
    const mockSetLanguageChangeUrls = jest.fn()

    const disclaimerPage = shallow(
      <DisclaimerPage languages={languages}
                      location={location}
                      disclaimer={disclaimer}
                      setLanguageChangeUrls={mockSetLanguageChangeUrls} />
    ).instance()
    expect(disclaimerPage.mapLanguageToUrl('en')).toBe('/augsburg/en/disclaimer')
  })

  const mockStore = configureMockStore([thunk])

  describe('connect', () => {
    test('should map state to props', () => {
      const store = new Store()
      store.init({
        disclaimer: new Payload(false),
        languages: new Payload(false),
        router: {params: {location: location}}
      })

      const tree = mount(
        <Provider store={store.redux}>
          <ConnectedDisclaimerPage />
        </Provider>
      )

      const disclaimerPageProps = tree.find(ConnectedDisclaimerPage).childAt(0).props()

      // todo add disclaimer and languages, WEBAPP-167
      expect(disclaimerPageProps).toEqual({
        location: 'augsburg',
        setLanguageChangeUrls: expect.any(Function)
      })
    })

    test('should map dispatch to props', () => {
      const store = mockStore({
        disclaimer: new Payload(false),
        languages: new Payload(false),
        router: {params: {location: location}}
      })

      const mapLanguageToUrl = (language) => language

      const testUrls = {
        en: 'en',
        ar: 'ar',
        de: 'de'
      }

      const tree = mount(
        <Provider store={store}>
          <ConnectedDisclaimerPage />
        </Provider>
      )

      // todo expect setLanguageChangeUrls action to be in store, but as we don't get events and languages from our
      // mocked endpoint no action is dispatched, WEBAPP-167

      const disclaimerPageProps = tree.find(ConnectedDisclaimerPage).childAt(0).props()

      let countActions = store.getActions().length

      disclaimerPageProps.setLanguageChangeUrls(mapLanguageToUrl, languages)
      expect(store.getActions()).toHaveLength(countActions + 1)

      expect(store.getActions()).toContainEqual({
        payload: testUrls,
        type: 'SET_LANGUAGE_CHANGE_URLS'
      })
    })
  })
})
