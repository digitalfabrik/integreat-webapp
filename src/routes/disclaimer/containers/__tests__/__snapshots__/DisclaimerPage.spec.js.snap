// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisclaimerPage should render 1`] = `
<Page
  content="this is a test content"
  title="Feedback, Kontakt und mÃ¶gliches Engagement"
/>
`;
